packages/language-server/src/plugins/yaml.ts

```diff
@@ -0,0 +1,144 @@
+import { Diagnostic, DiagnosticSeverity, MarkupContent, Range } from '@volar/language-server';
+import type { LanguageServicePlugin } from '@volar/language-service';
+import type { Provide } from 'volar-service-yaml';
+import { create as createYAMLService } from 'volar-service-yaml';
+import { URI, Utils } from 'vscode-uri';
+import type { CollectionConfig } from '../core/frontmatterHolders.js';
+import {
+	FrontmatterHolder,
+	SUPPORTED_FRONTMATTER_EXTENSIONS_KEYS,
+} from '../core/frontmatterHolders.js';
+
+export const create = (collectionConfigs: CollectionConfig[]): LanguageServicePlugin => {
+	const yamlPlugin = createYAMLService({
+		getLanguageSettings(context) {
+			if (!context.env.fs) {
+				return {};
+			}
+
+			const schemas = collectionConfigs.flatMap((workspaceCollectionConfig) => {
+				return workspaceCollectionConfig.config.collections.flatMap((collection) => {
+					return {
+						priority: 3,
+						fileMatch: SUPPORTED_FRONTMATTER_EXTENSIONS_KEYS.map(
+							(ext) => `volar-embedded-content://yaml_frontmatter_${collection.name}/**/*${ext}`,
+						),
+						uri: Utils.joinPath(
+							workspaceCollectionConfig.folder,
+							'.astro/collections',
+							`${collection.name}.schema.json`,
+						).toString(),
+					};
+				});
+			});
+
+			return {
+				completion: true,
+				format: false,
+				hover: true,
+				validate: true,
+				customTags: [],
+				yamlVersion: '1.2',
+				isKubernetes: false,
+				parentSkeletonSelectedFirst: false,
+				disableDefaultProperties: false,
+				schemas: schemas,
+			};
+		},
+	}) as LanguageServicePlugin<Provide>;
+
+	return {
+		...yamlPlugin,
+		capabilities: {
+			...yamlPlugin.capabilities,
+			codeLensProvider: undefined,
+		},
+		create(context) {
+			const yamlPluginInstance = yamlPlugin.create(context);
+
+			const languageService = yamlPluginInstance.provide!['yaml/languageService']();
+
+			if (context.env.onDidChangeWatchedFiles) {
+				context.env.onDidChangeWatchedFiles(async (events) => {
+					let hasChanges = false;
+
+					for (const change of events.changes) {
+						if (!change.uri.endsWith('.schema.json')) return;
+
+						if (languageService.resetSchema(change.uri)) {
+							hasChanges = true;
+						}
+					}
+
+					if (hasChanges) {
+						// TODO: Figure out how to refresh the diagnostics
+					}
+				});
+			}
+
+			return {
+				...yamlPluginInstance,
+				// Disable codelenses, we'll provide our own
```

Comment:
Not included in this PR, but disabling the built-in code lenses is still valuable, they're irrelevant for Astro users.